File: app/src/main/java/com/naukripatra/emicalculator/util/InterstitialAdManager.java
package com.naukripatra.emicalculator.util;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;

import com.google.android.gms.ads.AdError;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.FullScreenContentCallback;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.interstitial.InterstitialAd;
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback;

/**
 * Shows a full-screen interstitial the 1st time the user taps Calculate, then again every
 * 3rd time after (taps 1, 3, 6, 9, ...). The ad is preloaded ahead of time so most taps don't
 * wait on a network call, and {@code onComplete} always fires — whether or not an ad was shown —
 * so callers can safely navigate on afterward without ever blocking the user.
 */
public final class InterstitialAdManager {

    private static final String PREFS_NAME = "ad_prefs";
    private static final String KEY_CLICK_COUNT = "calculate_click_count";

    private static InterstitialAd interstitialAd;
    private static boolean loading;

    private InterstitialAdManager() {
    }

    public interface OnCompleteListener {
        void onComplete();
    }

    public static void preload(Context context) {
        if (interstitialAd != null || loading) {
            return;
        }
        loading = true;
        InterstitialAd.load(context.getApplicationContext(), AdConfig.INTERSTITIAL_AD_UNIT_ID,
                new AdRequest.Builder().build(), new InterstitialAdLoadCallback() {
                    @Override
                    public void onAdLoaded(@NonNull InterstitialAd ad) {
                        interstitialAd = ad;
                        loading = false;
                    }

                    @Override
                    public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                        interstitialAd = null;
                        loading = false;
                    }
                });
    }

    public static void showIfEligible(Activity activity, OnCompleteListener listener) {
        SharedPreferences prefs = activity.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        int count = prefs.getInt(KEY_CLICK_COUNT, 0) + 1;
        prefs.edit().putInt(KEY_CLICK_COUNT, count).apply();

        boolean eligible = count == 1 || count % 3 == 0;

        if (!eligible || interstitialAd == null) {
            listener.onComplete();
            preload(activity);
            return;
        }

        InterstitialAd adToShow = interstitialAd;
        interstitialAd = null;
        adToShow.setFullScreenContentCallback(new FullScreenContentCallback() {
            @Override
            public void onAdDismissedFullScreenContent() {
                preload(activity);
                listener.onComplete();
            }

            @Override
            public void onAdFailedToShowFullScreenContent(@NonNull AdError adError) {
                preload(activity);
                listener.onComplete();
            }
        });
        adToShow.show(activity);
    }
}

File: app/src/main/java/com/questionpapermaker/app/ads/InterstitialAdManager.kt
package com.questionpapermaker.app.ads

import android.app.Activity
import android.content.Context
import android.util.Log
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback

/**
 * Loads one interstitial ahead of time and shows it before a download completes. The user's
 * actual task (getting their PDF) is never blocked on the ad: if nothing is loaded yet or the
 * ad fails for any reason, [showThenRun] just runs [onComplete] immediately.
 */
class InterstitialAdManager(private val appContext: Context) {

    private var interstitialAd: InterstitialAd? = null
    private var isLoading = false

    fun preload() {
        if (interstitialAd != null || isLoading) return
        isLoading = true
        InterstitialAd.load(
            appContext,
            AdConfig.INTERSTITIAL_AD_UNIT_ID,
            AdRequest.Builder().build(),
            object : InterstitialAdLoadCallback() {
                override fun onAdLoaded(ad: InterstitialAd) {
                    isLoading = false
                    interstitialAd = ad
                }

                override fun onAdFailedToLoad(error: LoadAdError) {
                    isLoading = false
                    interstitialAd = null
                    Log.d(TAG, "Interstitial failed to load: ${error.message}")
                }
            }
        )
    }

    /** Shows the preloaded interstitial (if any) on [activity], then always calls [onComplete] exactly once. */
    fun showThenRun(activity: Activity, onComplete: () -> Unit) {
        val ad = interstitialAd
        if (ad == null) {
            preload()
            onComplete()
            return
        }

        ad.fullScreenContentCallback = object : FullScreenContentCallback() {
            override fun onAdDismissedFullScreenContent() {
                interstitialAd = null
                preload()
                onComplete()
            }

            override fun onAdFailedToShowFullScreenContent(adError: AdError) {
                interstitialAd = null
                preload()
                onComplete()
            }
        }
        ad.show(activity)
    }

    private companion object {
        const val TAG = "InterstitialAdManager"
    }
}
